cmd/thanos/rule.go

```diff
@@ -319,25 +325,52 @@ func runRule(
 		// Discover and resolve query addresses.
 		addDiscoveryGroups(g, queryClient, conf.query.dnsSDInterval)
 	}
+	var (
+		appendable storage.Appendable
+		queryable  storage.Queryable
+		db         *tsdb.DB
+	)
 
-	db, err := tsdb.Open(conf.dataDir, log.With(logger, "component", "tsdb"), reg, tsdbOpts, nil)
+	rwCfgYAML, err := conf.rwConfig.Content()
 	if err != nil {
-		return errors.Wrap(err, "open TSDB")
+		return err
 	}
 
-	level.Debug(logger).Log("msg", "removing storage lock file if any")
-	if err := removeLockfileIfAny(logger, conf.dataDir); err != nil {
-		return errors.Wrap(err, "remove storage lock files")
-	}
+	if len(rwCfgYAML) > 0 {
+		var rwCfg remotewrite.Config
+		rwCfg, err = remotewrite.LoadRemoteWriteConfig(rwCfgYAML)
+		if err != nil {
+			return err
+		}
+		walDir := filepath.Join(conf.dataDir, rwCfg.Name)
+		remoteStore, err := remotewrite.NewFanoutStorage(logger, reg, walDir, rwCfg)
+		if err != nil {
+			return errors.Wrap(err, "set up remote-write store for ruler")
+		}
+		appendable = remoteStore
+		queryable = remoteStore
+	} else {
+		db, err := tsdb.Open(conf.dataDir, log.With(logger, "component", "tsdb"), reg, tsdbOpts, nil)
```

Comment:
We can use `=` instead of `:=` here

Reply:
`tsdbStore` is used in creating the Store API server for ruler and the way I see it, running ruler in "stateless" mode means it's no longer a Store - maybe it makes sense to only create `tsdbStore` if `db` is not nil?

Reply:
Oh, that makes sense. If we don't want to consider it as store, let's check db and also stop grpc server is remote write is enabled.